File: Game2D/Engine/Boss_Hand1.cs
using System.Windows.Controls;
using System.Windows.Media.Imaging;
using System;

namespace Game2D.Engine
{
    public class Boss_Hand1 : GameObject
    {
        private Hero _hero;
        private double speed = 1.2;
        private int damage = 8;
        private int health = 120;
        private int moveType = 0;
        private int direction = 1;
        private double patrolRadius = 80;
        private double patrolCenterX, patrolCenterY;
        private double patrolAngle = 0;

        public bool IsDead => health <= 0;

        public Boss_Hand1(double x, double y, Hero hero)
        {
            X = x;
            Y = y;
            _hero = hero;
            patrolCenterX = x;
            patrolCenterY = y;
            
            Sprite = new Image
            {
                Source = new BitmapImage(new Uri("pack://application:,,,/Project/images/boss_hand1.png")),
                Width = 120,
                Height = 120
            };
        }

        public void SetDamage(int coef)
        {
            damage = coef;
        }

        public void SetSpeed(int speed)
        {
            this.speed = speed;
        }

        public void SetMoveType(int type)
        {
            moveType = type;
        }

        public void SetDirection(int dir)
        {
            direction = dir;
        }

        public override void Update()
        {
            if (_hero == null || !_hero.IsAlive) return;

            MovePattern();
            CheckOutOfBounds();
            CheckHeroCollision();
        }

        public void MovePattern()
        {
            switch (moveType)
            {
                case 0: 
                    ChaseHero();
                    break;
                case 1: 
                    PatrolMovement();
                    break;
                case 2: 
                    AttackMovement();
                    break;
            }
        }

        private void ChaseHero()
        {
            double dx = _hero.X - X;
            double dy = _hero.Y - Y;
            double dist = Math.Sqrt(dx * dx + dy * dy);
            if (dist > 1)
            {
                X += dx / dist * speed;
                Y += dy / dist * speed;
            }
        }

        private void PatrolMovement()
        {
            patrolAngle += 0.02 * direction;
            X = patrolCenterX + Math.Cos(patrolAngle) * patrolRadius;
            Y = patrolCenterY + Math.Sin(patrolAngle) * patrolRadius;
        }

        private void AttackMovement()
        {
            double dx = _hero.X - X;
            double dy = _hero.Y - Y;
            double dist = Math.Sqrt(dx * dx + dy * dy);
            if (dist > 1)
            {
                X += dx / dist * speed * 2;
                Y += dy / dist * speed * 2;
            }
        }

        public bool CheckOutOfBounds()
        {
            var world = MainWindow.CurrentGameWorld;
            if (world == null) return false;

            double maxX = world._canvas.ActualWidth - Sprite.Width;
            double maxY = world._canvas.ActualHeight - Sprite.Height;

            if (X < 0 || Y < 0 || X > maxX || Y > maxY)
            {
                X = Math.Max(0, Math.Min(X, maxX));
                Y = Math.Max(0, Math.Min(Y, maxY));
                return true;
            }
            return false;
        }

        public void CheckHeroCollision()
        {
            if (_hero != null && this.GetBounds().IntersectsWith(_hero.GetBounds()))
            {
                _hero.Health -= damage;
            }
        }

        public Hero GetNearestHero()
        {
            return _hero;
        }

        public void TakeDamage(int dmg)
        {
            health -= dmg;
            if (health <= 0)
            {
                if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                IsActive = false;
            }
        }
    }
} 

File: Game2D/Project/Boss_Hand1.java
import greenfoot.*;

public class Boss_Hand1 extends Actor {
    private int direction = 1;       // Направление движения (+1 или -1)
    private int moveType = 0;        // Тип движения (0: вертикаль, 1: горизонталь, 2: диагональ1, 3: диагональ2)
    private int speed = 1;           // Скорость движения
    private boolean hasHitHero = false; // Флаг, была ли атака по герою
    private int collisionRadius = 30; // Радиус обнаружения героя
    private int damage = 100; 
     
    public void set_damage(int coef){
        this.damage = damage + coef;
    } 
    
    public void act() {
        if (getWorld() == null) return; // Если рука удалена - выходим
        
        movePattern(); // Двигаем руку
        
        // Проверяем выход за границы (если true - рука удалена)
        if (checkOutOfBounds()) return;
        
        // Если ещё не били героя - проверяем коллизию
        if (!hasHitHero) {
            checkHeroCollision();
        }
    }

    // Установка скорости
    public void setSpeed(int speed) {
        this.speed = speed;
    }

    // Установка типа движения
    public void setMoveType(int type) {
        this.moveType = type;
    }

    // Установка направления
    public void setDirection(int dir) {
        this.direction = dir;
    }

    // Паттерны движения
    private void movePattern() {
        switch (moveType) {
            case 0: // вертикаль
                setLocation(getX(), getY() + direction * speed);
                break;
            case 1: // горизонталь
                setLocation(getX() + direction * speed, getY());
                break;
            case 2: // диагональ (из левого верхнего в правый нижний)
                setLocation(getX() + direction * speed, getY() + direction * speed);
                break;
            case 3: // диагональ (из правого верхнего в левый нижний)
                setLocation(getX() - direction * speed, getY() + direction * speed);
                break;
        }
    }

    // Проверка выхода за границы мира
    private boolean checkOutOfBounds() {
        World world = getWorld();
        if (world == null) return true;

        int x = getX();
        int y = getY();
        int width = world.getWidth();
        int height = world.getHeight();

        boolean outOfBounds = false;
        
        switch (moveType) {
            case 0: // вертикаль
                outOfBounds = (y <= 0 || y >= height - 1);
                break;
            case 1: // горизонталь
                outOfBounds = (x <= 0 || x >= width - 1);
                break;
            case 2: // диагональ1
            case 3: // диагональ2
                outOfBounds = (y <= 0 || y >= height - 1);
                break;
        }
        
        if (outOfBounds) {
            world.removeObject(this);
            return true;
        }
        return false;
    }

  private void checkHeroCollision() {
    if (getWorld() == null) return;

    java.util.List<Hero> heroes = getWorld().getObjects(Hero.class);
    for (Hero hero : heroes) {
        if (intersects(hero)) {
            hero.takeDamage(damage);
            hasHitHero = true;
            return;
        }
    }
}


    // Поиск ближайшего героя в радиусе
    private Hero getNearestHero() {
        if (getWorld() == null) return null;
        
        // Ищем всех героев в заданном радиусе
        java.util.List<Hero> heroes = getObjectsInRange(collisionRadius, Hero.class);
        if (!heroes.isEmpty()) {
            return heroes.get(0); // Возвращаем первого найденного героя
        }
        return null;
    }
}
